File: q1-25/q6.py
# 2020.07.10

class Solution:
    def convert(self, s: str, numRows: int) -> str:
        # initialize answer string
        answer = ""
        
        # check empty input
        if s == "":
            return ""
        
        # calculate how many element are in each group
        # group means patterned element group in shape
        # ex: a     g
        #     b   f h   l
        #     c e   i k
        #     d     j
        # then abcdef is a group
        group_size = numRows * 2 - 2
        
        if group_size == 0:
            return s
        
        # calculate total number of groups
        if len(s) % group_size == 0:
            group_num_total = len(s) // group_size
        else:
            group_num_total = len(s) // group_size + 1

        # for each element, store which group it's in and its position in its group
        group_num = []
        group_index = []
        
        for i in range(0, len(s)):
            group_num.append(i // group_size)
            group_index.append(i % group_size)

        # index should go before group
        # from the first group to the last
        for index in range(0, group_size//2 + 1):
            # from the first index to the middle
            for group in range(0, group_num_total):
                # get the element to store
                for i in range(0, len(s)):
                    if group_num[i] == group and group_index[i] == index:
                        answer = answer + s[i]
                        breaks
                # consider the cases when 2 elements are on the same row
                # store the element as well
                if index != 0 and index != group_size//2:
                    for i in range(0, len(s)):
                        if group_num[i] == group and group_index[i] + index == group_size:
                            answer = answer + s[i]
                            break
        
        return answer

File: Amazon/q6.java
// 2022.06.29
// Problem Statement:
// https://leetcode.com/problems/zigzag-conversion/

// idea: group by 2*numRows-2 elements, iteration ordered by rows
// in row 0 and numRows-1, only 1 char in each group should be added
// for rows in the middle, add 2 chars in each group
class Solution {
    public String convert(String s, int numRows) {
        String answer = "";
        if (numRows==1) return s;
        for (int i=0; i<numRows; i++) {
            for (int j=0; j<s.length()/(2*numRows-2)+1; j++) { // iterate over all groups
                if (i==0) {
                    if (j*(2*numRows-2) < s.length()) { // check boundary
                        answer = answer + s.charAt(j*(2*numRows-2));
                    }
                } else if (i==numRows-1) {
                    if (j*(2*numRows-2)+numRows-1 < s.length()) {
                        answer = answer + s.charAt(j*(2*numRows-2)+numRows-1);
                    }
                } else {
                    if (j*(2*numRows-2)+i < s.length()) {
                        answer = answer + s.charAt(j*(2*numRows-2)+i);
                    }
                    if (j*(2*numRows-2)+(2*numRows-2-i) < s.length()) {
                        answer = answer + s.charAt(j*(2*numRows-2)+(2*numRows-2-i));
                    }
                }
            }
        }
        return answer;
    }
}